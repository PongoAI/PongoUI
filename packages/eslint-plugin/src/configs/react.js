/** @type {import("eslint").Linter.Config} */
const config = {
  extends: ['airbnb-typescript', 'prettier'],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: 'tsconfig.json',
  },
  plugins: ['deprecation', 'import', '@typescript-eslint', 'jest', 'jsx-a11y', 'react', 'react-hooks'],
  settings: {
    'import/parsers': {
      '@typescript-eslint/parser': ['.ts', '.tsx'],
    },
    'import/resolver': {
      typescript: {
        alwaysTryTypes: true,
      },
    },
    'no-unused-expressions': 'off',
  },
  env: {
    browser: true,
    es6: true,
    node: true,
  },

  ignorePatterns: ['coverage', 'dist', 'dist-storybook', 'lib', 'lib-commonjs', '**/__snapshots__', 'node_modules'],
  rules: {
    'no-unused-vars': 'error',
    'no-use-before-define': 'off',
    '@typescript-eslint/no-use-before-define': ['error'],
    'import/no-extraneous-dependencies': ['error', { devDependencies: true }],
    'import/extensions': [
      'error',
      'ignorePackages',
      {
        js: 'never',
        jsx: 'never',
        ts: 'never',
        tsx: 'never',
      },
    ],
  },
};

module.exports = config;
