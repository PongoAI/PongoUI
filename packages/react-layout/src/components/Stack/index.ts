export * from './Stack';
export * from './useStackStyles';
export * from './Stack.types';
export * from './useStack';
