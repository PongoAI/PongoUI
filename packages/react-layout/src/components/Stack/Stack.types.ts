import * as React from 'react';
import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type StackSlots = {
  /**
   * The root of the Stack.
   */
  root: Slot<'div'>;
};

export interface StackCommons {
  /**
   * Whether the stack should be rendered vertically.
   *
   * @defaultValue true
   */
  vertical?: boolean;

  /**
   * The divider to render between each child element.
   */
  divider?: React.ReactNode;

  /**
   * The appearance of the Stack.
   * Focused has a lower padding and gap size then relaxed.
   * Relaxed should be used when the content of your app is minimal.
   *
   * @defaultValue focused
   */
  appearance?: 'focused' | 'relaxed';

  /**
   * How the components should be aligned horizontally.
   */
  horizontalAlignment?: 'start' | 'center' | 'end';

  /**
   * How the component should be aligned vertically.
   */
  verticalAlignment?: 'start' | 'center' | 'end';

  /**
   * Whether the content should be inline. This will cause the content to grow based on its children's size rather than parent.
   */
  inline?: boolean;

  /**
   * Whether the content should wrap.
   */
  wrap?: boolean;

  /**
   * Whether the Stack should grow to fit its container.
   */
  grow?: boolean;
}

export type StackProps = ComponentProps<StackSlots> & StackCommons;

export type StackState = ComponentState<StackSlots> & StackCommons;
