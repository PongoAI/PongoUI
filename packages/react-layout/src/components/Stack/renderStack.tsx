import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { StackState, StackSlots } from './Stack.types';

export const renderStack = (state: StackState) => {
  const { slots, slotProps } = getSlots<StackSlots>(state);

  return <slots.root {...slotProps.root} />;
};
