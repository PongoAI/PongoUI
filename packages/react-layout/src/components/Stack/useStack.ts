import * as React from 'react';
import { getNativeElementProps } from '@fluentui/react-utilities';
import type { StackProps, StackState } from './Stack.types';

export const useStack = (props: StackProps, ref: React.Ref<HTMLElement>): StackState => {
  const {
    vertical = false,
    divider,
    appearance = 'focused',
    horizontalAlignment = 'center',
    verticalAlignment = 'center',
    grow = false,
    inline = false,
    wrap = true,
  } = props;

  const state: StackState = {
    vertical,
    divider,
    appearance,
    inline,
    wrap,
    grow,
    horizontalAlignment,
    verticalAlignment,
    components: {
      root: 'div',
    },
    root: getNativeElementProps('div', {
      ref,
      ...props,
    }),
  };

  if (divider) {
    const oldChildren: any = state.root.children;
    const newChildren = [];

    if (oldChildren && oldChildren?.length > 1) {
      for (let i = 0; i < oldChildren.length - 1; i++) {
        newChildren.push(oldChildren[i]);
        newChildren.push(divider);
      }
      newChildren.push(oldChildren[oldChildren.length - 1]);
    }
    state.root.children = newChildren;
  }

  return state;
};
