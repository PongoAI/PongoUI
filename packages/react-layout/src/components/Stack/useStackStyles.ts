import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type { StackState } from './Stack.types';
import { tokens } from '@pongo-ui/react-theme';

const useStyles = makeStyles({
  grow: {
    width: '100%',
    height: '100%',
    ['> *' as any]: {
      flexGrow: 1,
    },
  },

  inline: {
    display: 'inline-flex',
  },

  flex: {
    display: 'flex',
  },

  horizontal: {
    flexDirection: 'row',
  },

  vertical: {
    flexDirection: 'column',
  },

  centerHorizontal: {
    alignItems: 'center',
  },

  startHorizontal: {
    alignItems: 'start',
  },

  endHorizontal: {
    alignItems: 'end',
  },

  centerVertical: {
    justifyContent: 'center',
  },

  startVertical: {
    justifyContent: 'start',
  },

  endVertical: {
    justifyContent: 'end',
  },

  wrap: {
    flexWrap: 'wrap',
  },

  focused: {
    ...shorthands.gap(tokens.focusedGap),
  },

  relaxed: {
    ...shorthands.gap(tokens.relaxedGap),
  },
});

export const useStackStyles = (state: StackState) => {
  const styles = useStyles();

  state.root.className = mergeClasses(
    state.vertical ? styles.vertical : styles.horizontal,
    state.inline ? styles.inline : styles.flex,
    state.wrap && styles.wrap,
    state.grow && styles.grow,
    styles[state.appearance!],
    state.vertical
      ? styles[(state.horizontalAlignment! + 'Horizontal') as 'endHorizontal' | 'startHorizontal' | 'centerHorizontal']
      : styles[(state.horizontalAlignment! + 'Vertical') as 'endVertical' | 'startVertical' | 'centerVertical'],
    state.vertical
      ? styles[(state.verticalAlignment! + 'Vertical') as 'endVertical' | 'startVertical' | 'centerVertical']
      : styles[(state.verticalAlignment! + 'Horizontal') as 'endHorizontal' | 'startHorizontal' | 'centerHorizontal'],
    state.root.className,
  );

  return state;
};
