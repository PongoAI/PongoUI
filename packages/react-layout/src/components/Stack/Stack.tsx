import * as React from 'react';
import { useStack } from './useStack';
import { useStackStyles } from './useStackStyles';
import { renderStack } from './renderStack';
import type { StackProps } from './Stack.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * The Stack helps position and handle padding for a series of components.
 */
export const Stack: ForwardRefComponent<StackProps> = React.forwardRef((props, ref) => {
  const state = useStack(props, ref);

  useStackStyles(state);

  return renderStack(state);
});
Stack.displayName = 'Stack';
