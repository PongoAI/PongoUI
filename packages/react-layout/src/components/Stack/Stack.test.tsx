import React from 'react';
import { Stack } from './Stack';
import { isConformant } from '../../common/isConformant';
import { render } from '@testing-library/react';

describe('Input', () => {
  isConformant({
    Component: Stack,
    displayName: 'Stack',
  });

  it('renders a basic Stack', () => {
    const renderedComponent = render(<Stack />);

    expect(renderedComponent.baseElement).toMatchSnapshot();
  });

  it('renders a divider Stack', () => {
    const renderedComponent = render(
      <Stack divider={' - '}>
        <p>Hello</p>
        <p>Hello</p>
      </Stack>,
    );

    expect(renderedComponent.baseElement).toMatchSnapshot();
  });

  it('renders a divider Stack with once child', () => {
    const renderedComponent = render(
      <Stack divider={' - '}>
        <p>Hello</p>
      </Stack>,
    );

    expect(renderedComponent.baseElement).toMatchSnapshot();
  });

  it('renders a vertical Stack', () => {
    const renderedComponent = render(<Stack vertical />);

    expect(renderedComponent.baseElement).toMatchSnapshot();
  });

  it('Correctly applies horizontal/vertical alignment to Stack (vertical)', () => {
    const renderedComponent = render(<Stack vertical horizontalAlignment="center" verticalAlignment="center" />);

    expect(renderedComponent.baseElement).toMatchSnapshot();
  });

  it('Correctly applies horizontal/vertical alignment to Stack (horizontal)', () => {
    const renderedComponent = render(<Stack horizontalAlignment="center" verticalAlignment="center" />);

    expect(renderedComponent.baseElement).toMatchSnapshot();
  });
});
