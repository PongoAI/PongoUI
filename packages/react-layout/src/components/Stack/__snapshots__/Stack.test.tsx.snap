// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack Correctly applies horizontal/vertical alignment to Stack (horizontal) 1`] = `
<body>
  <div>
    <div
      class="___1n7lss6 f1063pyq f22iagw fni485r fnmxc1e fynvm4a f4d9j23 f122n59"
    />
  </div>
</body>
`;

exports[`Stack Correctly applies horizontal/vertical alignment to Stack (vertical) 1`] = `
<body>
  <div>
    <div
      class="___1dkafrz f1vx9l62 f22iagw fni485r fnmxc1e fynvm4a f122n59 f4d9j23"
    />
  </div>
</body>
`;

exports[`Stack renders a basic Stack 1`] = `
<body>
  <div>
    <div
      class="___2q39d80 f1063pyq f22iagw fni485r fnmxc1e fynvm4a f1s9ku6b f1s2louj"
    />
  </div>
</body>
`;

exports[`Stack renders a divider Stack 1`] = `
<body>
  <div>
    <div
      class="___2q39d80 f1063pyq f22iagw fni485r fnmxc1e fynvm4a f1s9ku6b f1s2louj"
    >
      <p>
        Hello
      </p>
       - 
      <p>
        Hello
      </p>
    </div>
  </div>
</body>
`;

exports[`Stack renders a divider Stack with once child 1`] = `
<body>
  <div>
    <div
      class="___2q39d80 f1063pyq f22iagw fni485r fnmxc1e fynvm4a f1s9ku6b f1s2louj"
    />
  </div>
</body>
`;

exports[`Stack renders a vertical Stack 1`] = `
<body>
  <div>
    <div
      class="___17t8msf f1vx9l62 f22iagw fni485r fnmxc1e fynvm4a f1s2louj f1s9ku6b"
    />
  </div>
</body>
`;
